shpkpr/deployment/bluegreen/deployment.py

```diff
@@ -0,0 +1,148 @@
+# stdlib imports
+import logging
+import time
+
+# local imports
+from .prepare import prepare_app_definition
+from .validate import Validator
+from .wait import SwapApplicationTimeout
+from .wait import Waiter
+from shpkpr.marathon import DeploymentFailed
+
+
+logger = logging.getLogger(__name__)
+
+
+class BlueGreenDeployment(object):
+    """BluegreenDeployment implements a blue-green deployment strategy for
+    web-facing applications on Marathon.
+
+    It builds on the primitives provided by Marathon to provide zero-downtime
+    deployments for HTTP services exposed to the web via Marathon-LB.
+    """
+
+    def __init__(self, marathon_client, marathon_lb_client, timeout, app_definitions):
+        self.marathon_client = marathon_client
+        self.marathon_lb_client = marathon_lb_client
+        self.timeout = timeout
+        self.app_definitions = app_definitions
+
+    def execute(self, force=False):
+        """Execute a bluegreen deployment.
+
+        The mechanics of a bluegreen deploy mean that we have to deploy each
+        application in turn (instead of the all-at-once approach used for
+        standard deployments).
+
+        NOTE: We could maybe be smarter here in future e.g. perhaps don't tear
+        down old stacks until _all_ new stacks are running so we can roll back
+        partially successful invocations.
+        """
+        for app_definition in self.app_definitions:
+            self.execute_for_app(app_definition, force)
+
+    def execute_for_app(self, app_definition, force):
+        """Execute a bluegreen deployment for a single application.
+        """
+        deployment = BlueGreenDeploymentSingleApp(self.marathon_client,
+                                                  self.marathon_lb_client,
+                                                  self.timeout,
+                                                  app_definition)
+        return deployment.execute(force=force)
+
+
+class BlueGreenDeploymentSingleApp(object):
+
+    def __init__(self, marathon_client, marathon_lb_client, timeout, app_definition):
+        self.marathon_client = marathon_client
+        self.marathon_lb_client = marathon_lb_client
+        self.timeout = timeout
+        self.app_definition = app_definition
+
+    def execute(self, force=False):
```

Comment:
Yeah, it isn't right now. It never really made sense the way it was implemented before (essentially all non-interactive bluegreen deploys required `force` to succeed) and I'm not sure I want to re-add it without solving those flow issues. It's mainly here so as to keep the API between `StandradDeployment` and `BluegreenDeployment` consistent for now.